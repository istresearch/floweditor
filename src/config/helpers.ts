import {
  FlowTypes,
  FlowTypeVisibility,
  Operator,
  Type,
  VISIBILITY_ONLINE
} from 'config/interfaces';
import { FlowEditorConfig } from 'flowTypes';

export const isOnlineFlowType = (flowType: FlowTypes) => {
  return !!VISIBILITY_ONLINE.find((type: FlowTypes) => type === flowType);
};

export const filterOperators = (operators: Operator[], config: FlowEditorConfig): Operator[] => {
  return filterVisibility(operators, config);
};

export const filterTypeConfigs = (typeConfigs: Type[], config: FlowEditorConfig): Type[] => {
  return filterVisibility(typeConfigs, config);
};

const filterVisibility = (items: FlowTypeVisibility[], config: FlowEditorConfig): any[] => {
  return items.filter((item: FlowTypeVisibility) => {
    // if we have a filter on our type, don't return it unless its present in our config
    if (item.filter) {
      if (!(config.filters || []).find((name: string) => name === item.filter)) {
        return false;
      }
    }

    if (item.visibility === undefined) {
      return true;
    }

    return item.visibility.findIndex((ft: FlowTypes) => ft === config.flowType) > -1;
  });
};
