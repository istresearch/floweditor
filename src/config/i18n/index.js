import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import LanguageDetector from 'i18next-browser-languagedetector';

// TODO: consider chunking out translation files
// and using something like XHR backend to lazily
// load only the required language.
// See https://www.i18next.com/how-to/add-or-load-translations

import defaults from './defaults.json';
import cs from './cs/resource.json';
import es from './es/resource.json';
import fr from './fr/resource.json';
import mn from './mn/resource.json';
import ptBR from './pt-br/resource.json';

i18n
  .use(LanguageDetector)
  .use(initReactI18next)
  .init({
    detection: {
      order: ['querystring', 'htmlTag']
    },
    resources: {
      defaults: { translation: defaults },
      cs: { translation: cs },
      es: { translation: es },
      fr: { translation: fr },
      mn: { translation: mn },
      'pt-br': { translation: ptBR }
    },
    fallbackLng: ['defaults'],
    interpolation: {
      prefix: '[[',
      suffix: ']]'
    },
    lowerCaseLng: true
  });

export default i18n;
