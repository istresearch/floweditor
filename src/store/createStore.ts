import { applyMiddleware, createStore, Middleware } from 'redux';
import { composeWithDevTools } from '@redux-devtools/extension';
import thunk from 'redux-thunk';

import rootReducer from 'store/rootReducer';
import AppState, { initialState } from 'store/state';

const middlewares: Middleware[] = [thunk];

export default (state: AppState = initialState) => {
  const store = createStore(
    rootReducer,
    state,
    composeWithDevTools(applyMiddleware(...middlewares))
  );

  /* istanbul ignore next */
  if (module.hot) {
    module.hot.accept('./rootReducer', () => {
      const { default: nextRootReducer } = require('./rootReducer');
      store.replaceReducer(nextRootReducer);
    });
  }

  return store;
};
