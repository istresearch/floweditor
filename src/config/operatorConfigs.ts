import {
  Operator,
  OperatorMap,
  Operators,
  VISIBILITY_ONLINE,
  VISIBILITY_HIDDEN
} from 'config/interfaces';
import i18n from 'config/i18n';

export const intentOperatorList: Operator[] = [
  {
    type: Operators.has_top_intent,
    verboseName: i18n.t('operators.has_top_intent', 'has top intent'),
    operands: 2,
    visibility: VISIBILITY_ONLINE
  },
  {
    type: Operators.has_intent,
    verboseName: i18n.t('operators.has_intent', 'has intent'),
    operands: 2,
    visibility: VISIBILITY_ONLINE
  }
];

export const operatorConfigList: Operator[] = [
  {
    type: Operators.has_any_word,
    verboseName: i18n.t('operators.has_any_word', 'has any of the words'),
    operands: 1
  },
  {
    type: Operators.has_all_words,
    verboseName: i18n.t('operators.has_all_words', 'has all of the words'),
    operands: 1
  },
  {
    type: Operators.has_phrase,
    verboseName: i18n.t('operators.has_phrase', 'has the phrase'),
    operands: 1
  },
  {
    type: Operators.has_only_phrase,
    verboseName: i18n.t('operators.has_only_phrase', 'has only the phrase'),
    operands: 1
  },
  {
    type: Operators.has_beginning,
    verboseName: i18n.t('operators.has_beginning', 'starts with'),
    operands: 1
  },
  {
    type: Operators.has_text,
    verboseName: i18n.t('operators.has_text', 'has some text'),
    operands: 0,
    categoryName: 'Has Text'
  },
  {
    type: Operators.has_number,
    verboseName: i18n.t('operators.has_number', 'has a number'),
    operands: 0,
    categoryName: 'Has Number'
  },
  {
    type: Operators.has_number_between,
    verboseName: i18n.t('operators.has_number_between', 'has a number between'),
    operands: 2
  },
  {
    type: Operators.has_number_lt,
    verboseName: i18n.t('operators.has_number_lt', 'has a number below'),
    operands: 1
  },
  {
    type: Operators.has_number_lte,
    verboseName: i18n.t('operators.has_number_lte', 'has a number at or below'),
    operands: 1
  },
  {
    type: Operators.has_number_eq,
    verboseName: i18n.t('operators.has_number_eq', 'has a number equal to'),
    operands: 1
  },
  {
    type: Operators.has_number_gte,
    verboseName: i18n.t('operators.has_number_gte', 'has a number at or above'),
    operands: 1
  },
  {
    type: Operators.has_number_gt,
    verboseName: i18n.t('operators.has_number_gt', 'has a number above'),
    operands: 1
  },
  {
    type: Operators.has_date,
    verboseName: i18n.t('operators.has_date', 'has a date'),
    operands: 0,
    categoryName: i18n.t('operators.has_date_category', 'Has Date')
  },
  {
    type: Operators.has_date_lt,
    verboseName: i18n.t('operators.has_date_lt', 'has a date before'),
    operands: 1
  },
  {
    type: Operators.has_date_eq,
    verboseName: i18n.t('operators.has_date_eq', 'has a date equal to'),
    operands: 1
  },
  {
    type: Operators.has_date_gt,
    verboseName: i18n.t('operators.has_date_gt', 'has a date after'),
    operands: 1
  },
  {
    type: Operators.has_time,
    verboseName: i18n.t('operators.has_time', 'has a time'),
    operands: 0,
    categoryName: 'Has Time'
  },
  {
    type: Operators.has_group,
    verboseName: i18n.t('operators.has_group', 'is in the group'),
    operands: 1,
    visibility: VISIBILITY_HIDDEN
  },
  {
    type: Operators.has_category,
    verboseName: i18n.t('operators.has_category', 'has the category'),
    operands: 0,
    visibility: VISIBILITY_HIDDEN
  },
  {
    type: Operators.has_phone,
    verboseName: i18n.t('operators.has_phone', 'has a phone number'),
    operands: 0,
    categoryName: i18n.t('operators.has_phone_category', 'Has Phone')
  },
  {
    type: Operators.has_email,
    verboseName: i18n.t('operators.has_email', 'has an email'),
    operands: 0,
    categoryName: i18n.t('operators.has_email_category', 'Has Email')
  },
  {
    type: Operators.has_state,
    verboseName: i18n.t('operators.has_state', 'has state'),
    operands: 0,
    categoryName: i18n.t('operators.has_state_category', 'Has State')
  },
  {
    type: Operators.has_district,
    verboseName: i18n.t('operators.has_district', 'has district'),
    operands: 1,
    categoryName: i18n.t('operators.has_district_category', 'Has District')
  },
  {
    type: Operators.has_ward,
    verboseName: i18n.t('operators.has_ward', 'has ward'),
    operands: 2,
    categoryName: i18n.t('operators.has_ward_category', 'Has Ward')
  },
  {
    type: Operators.has_error,
    verboseName: i18n.t('operators.has_error', 'has an error'),
    operands: 0,
    categoryName: i18n.t('operators.has_error_category', 'Has Error'),
    visibility: VISIBILITY_HIDDEN
  },
  {
    type: Operators.has_value,
    verboseName: i18n.t('operators.has_value', 'is not empty'),
    operands: 0,
    categoryName: i18n.t('operators.has_value_category', 'Not Empty'),
    visibility: VISIBILITY_HIDDEN
  },
  {
    type: Operators.has_pattern,
    verboseName: i18n.t('operators.has_pattern', 'matches regex'),
    operands: 1
  }
];

export const operatorConfigMap: OperatorMap = [...operatorConfigList, ...intentOperatorList].reduce(
  (map: OperatorMap, operatorConfig: Operator) => {
    map[operatorConfig.type] = operatorConfig;
    return map;
  },
  {}
);

/**
 * Shortcut for constant lookup of operator config in operator configs map
 * @param {string} type - The type of the operator config to return, e.g. 'send_msg'
 * @returns {Object} - The operator config found at operatorConfigs[type] or -1
 */
export const getOperatorConfig = (type: Operators): Operator => operatorConfigMap[type];
