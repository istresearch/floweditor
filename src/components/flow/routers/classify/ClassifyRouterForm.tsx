import { react as bindCallbacks } from 'auto-bind';
import Dialog, { ButtonSet, Tab } from 'components/dialog/Dialog';
import { hasErrors, renderIssues } from 'components/flow/actions/helpers';
import { RouterFormProps } from 'components/flow/props';
import { nodeToState, stateToNode } from './helpers';
import { createResultNameInput } from 'components/flow/routers/widgets';
import TypeList from 'components/nodeeditor/TypeList';
import * as React from 'react';
import { FormState, mergeForm, StringEntry, FormEntry } from 'store/nodeEditor';
import {
  Alphanumeric,
  Required,
  shouldRequireIf,
  StartIsNonNumeric,
  validate
} from 'store/validators';
import CaseList, { CaseProps } from 'components/flow/routers/caselist/CaseList';
import AssetSelector from 'components/form/assetselector/AssetSelector';
import { Asset } from 'store/flowContext';
import { renderIf } from 'utils';
import { intentOperatorList } from 'config/operatorConfigs';
import TextInputElement from 'components/form/textinput/TextInputElement';
import { DEFAULT_OPERAND } from 'components/nodeeditor/constants';
import { fetchAsset } from 'external';
import styles from './ClassifyRouterForm.module.scss';
import i18n from 'config/i18n';

export interface ClassifyRouterFormState extends FormState {
  hiddenCases: CaseProps[];
  resultName: StringEntry;
  classifier: FormEntry;
  cases: CaseProps[];
  operand: StringEntry;
}

export default class ClassifyRouterForm extends React.Component<
  RouterFormProps,
  ClassifyRouterFormState
> {
  constructor(props: RouterFormProps) {
    super(props);

    this.state = nodeToState(this.props.nodeSettings);
    bindCallbacks(this, {
      include: [/^handle/]
    });

    // we need to resolve our classifier for intent selection
    if (this.state.classifier.value) {
      // TODO: don't use asset as intermediary now that AssetSelector deals in native options
      fetchAsset(this.props.assetStore.classifiers, this.state.classifier.value.uuid).then(
        (classifier: Asset) => {
          if (classifier) {
            this.handleUpdate({
              classifier: { ...this.state.classifier.value, ...classifier.content }
            });
          }
        }
      );
    }
  }

  private handleUpdate(
    keys: {
      resultName?: string;
      classifier?: any;
    },
    submitting = false
  ): boolean {
    const updates: Partial<ClassifyRouterFormState> = {};

    if (keys.hasOwnProperty('resultName')) {
      updates.resultName = validate(i18n.t('forms.result_name', 'Result Name'), keys.resultName, [
        shouldRequireIf(submitting)
      ]);
    }

    if (keys.hasOwnProperty('classifier')) {
      updates.classifier = validate(i18n.t('forms.classifier', 'Classifier'), keys.classifier, [
        shouldRequireIf(submitting)
      ]);
    }

    const updated = mergeForm(this.state, updates);

    // update our form
    this.setState(updated);
    return updated.valid;
  }

  private handleCasesUpdated(cases: CaseProps[]): void {
    const invalidCase = cases.find((caseProps: CaseProps) => !caseProps.valid);
    this.setState({ cases, valid: !invalidCase });
  }

  private handleUpdateResultName(value: string): void {
    const resultName = validate(i18n.t('forms.result_name', 'Result Name'), value, [
      Required,
      Alphanumeric,
      StartIsNonNumeric
    ]);
    this.setState({
      resultName,
      valid: this.state.valid && !hasErrors(resultName)
    });
  }

  private handleSave(): void {
    // if we still have invalid cases, don't move forward
    const invalidCase = this.state.cases.find((caseProps: CaseProps) => !caseProps.valid);
    if (invalidCase) {
      return;
    }

    // validate our result name in case they haven't interacted
    const valid = this.handleUpdate(
      {
        resultName: this.state.resultName.value,
        classifier: this.state.classifier.value
      },
      true
    );

    if (valid) {
      this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
      this.props.onClose(false);
    }
  }

  private handleClassifierUpdated(selected: Asset[]): void {
    this.handleUpdate({ classifier: selected[0] });
  }

  private handleOperandUpdated(value: string): void {
    this.setState({
      operand: validate(i18n.t('forms.operand', 'Operand'), value, [Required])
    });
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.ok', 'Ok'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  private dialog: Dialog;

  private renderEdit(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    const tabs: Tab[] = [
      {
        name: 'Classifier Input',
        checked: this.state.operand.value !== DEFAULT_OPERAND,
        body: (
          <>
            <p>
              Enter an expression to use as the input to your classifier. To classify the last
              response from the contact use <code>{DEFAULT_OPERAND}</code>.
            </p>
            <TextInputElement
              name={i18n.t('forms.operand', 'Operand')}
              showLabel={false}
              autocomplete={true}
              onChange={this.handleOperandUpdated}
              entry={this.state.operand}
            />
          </>
        )
      }
    ];

    return (
      <Dialog
        title={typeConfig.name}
        headerClass={typeConfig.type}
        buttons={this.getButtons()}
        tabs={tabs}
        ref={ele => {
          this.dialog = ele;
        }}
      >
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        <p>
          <span>Run </span>
          <span
            className={styles.link}
            onClick={() => {
              this.dialog.showTab(0);
            }}
          >
            {this.state.operand.value === DEFAULT_OPERAND
              ? 'the last response'
              : this.state.operand.value}
          </span>
          <span> through the classifier...</span>
        </p>
        <AssetSelector
          key="select_classifier"
          name={i18n.t('forms.classifier', 'Classifier')}
          placeholder="Select the classifier to use"
          assets={this.props.assetStore.classifiers}
          onChange={this.handleClassifierUpdated}
          entry={this.state.classifier}
        />

        {renderIf(!!this.state.classifier.value)(
          <CaseList
            data-spec="cases"
            cases={this.state.cases}
            onCasesUpdated={this.handleCasesUpdated}
            operators={intentOperatorList}
            classifier={this.state.classifier.value}
          />
        )}

        {createResultNameInput(this.state.resultName, this.handleUpdateResultName)}
        {renderIssues(this.props)}
      </Dialog>
    );
  }

  public render(): JSX.Element {
    return this.renderEdit();
  }
}
