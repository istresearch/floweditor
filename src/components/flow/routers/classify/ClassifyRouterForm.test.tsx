import { Types } from 'config/interfaces';
import * as React from 'react';
import { AssetType, RenderNode } from 'store/flowContext';
import {
  fireEvent,
  render,
  getCallParams,
  fireTembaSelect,
  getByDisplayValue,
  fireChangeText,
  getByTestId
} from 'test/utils';
import { mock } from 'testUtils';
import { createClassifyRouter, getRouterFormProps } from 'testUtils/assetCreators';
import * as utils from 'utils';
import ClassifyRouterForm from './ClassifyRouterForm';
import * as external from 'external';

mock(
  external,
  'fetchAsset',
  utils.fetchAsset({
    id: 'purrington',
    name: 'Purrington',
    type: AssetType.Classifier,
    content: {
      intents: ['fact', 'opinion']
    }
  })
);

mock(utils, 'createUUID', utils.seededUUIDs());

const routerNode = createClassifyRouter();

const getProps = () => {
  const formProps = getRouterFormProps(routerNode);
  formProps.assetStore = { classifiers: { items: {}, type: AssetType.Classifier } };
  return formProps;
};

describe(ClassifyRouterForm.name, () => {
  it('should render', () => {
    const { baseElement } = render(<ClassifyRouterForm {...getProps()} />);
    expect(baseElement).toMatchSnapshot('default render');
  });

  it('defaults to correct result name', () => {
    const props = getProps();
    props.nodeSettings.originalNode.node.router.result_name = 'Initial Result Name';
    const { baseElement } = render(<ClassifyRouterForm {...props} />);
    expect(baseElement).toMatchSnapshot('initial result name');
  });

  it('updates the result name', () => {
    const { baseElement, getByTestId } = render(<ClassifyRouterForm {...getProps()} />);
    const result = getByTestId('Result Name');
    fireChangeText(result, 'Updated Result Name');
    expect(baseElement).toMatchSnapshot('updated result name');
  });
});
