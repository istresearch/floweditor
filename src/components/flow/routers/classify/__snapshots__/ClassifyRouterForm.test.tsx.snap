// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClassifyRouterForm defaults to correct result name: initial result name 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="tabs"
      >
        <div
          class="tab "
        >
          Classifier Input
           
          <span
            class="tab_icon fe-check"
          />
        </div>
      </div>
      <div
        class="header split_by_intent"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Split by Intent
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                errors="[]"
                expressions=""
                name="type"
                namekey="description"
                searchable="true"
                valuekey="type"
                values="[{\\"type\\":\\"split_by_intent\\",\\"name\\":\\"Split by Intent\\",\\"description\\":\\"Split by intent\\",\\"localizeableKeys\\":[\\"exits\\"],\\"aliases\\":[\\"call_classifier\\"],\\"visibility\\":[\\"messaging\\",\\"messaging_background\\",\\"voice\\"],\\"filter\\":\\"classifier\\"}]"
              />
            </div>
          </div>
        </div>
        <p>
          <span>
            Run 
          </span>
          <span
            class="link"
          >
            @input
          </span>
          <span>
             through the classifier...
          </span>
        </p>
        <div
          class="ele"
        >
          <div
            class="normal "
          >
            <temba-select
              cachekey="0"
              data-testid="temba_select_classifier"
              errors="[]"
              expressions=""
              name="Classifier"
              namekey="name"
              placeholder="Select the classifier to use"
              valuekey="uuid"
              values="[{\\"uuid\\":\\"purrington\\",\\"name\\":\\"Purrington\\"}]"
            />
          </div>
        </div>
        <div
          class="case_list_container "
          tabindex="0"
        >
          <div
            class="case_list"
          >
            <div
              class="kase case_list_case"
            >
              <div
                class="ele group"
              >
                <div
                  class="kase has_top_intent"
                  data-draggable="true"
                >
                  <span
                    class="fe-chevrons-expand dnd_icon"
                    data-draggable="true"
                  />
                  <div
                    class="choice"
                  >
                    <div
                      class="small "
                    >
                      <temba-select
                        data-testid="temba_select_operator"
                        errors="[]"
                        expressions=""
                        name="operator"
                        namekey="verboseName"
                        valuekey="type"
                        values="[{\\"type\\":\\"has_top_intent\\",\\"verboseName\\":\\"has top intent\\",\\"operands\\":2,\\"visibility\\":[\\"messaging\\",\\"messaging_background\\",\\"voice\\"]}]"
                      />
                    </div>
                  </div>
                  <div
                    class="multi_operand"
                  >
                    <div
                      style="width: 114px;"
                    >
                      <div
                        class="ele"
                      >
                        <div
                          class="small "
                        >
                          <temba-select
                            data-testid="temba_select_intent"
                            errors="[]"
                            expressions=""
                            hideerrors="true"
                            name="Intent"
                            namekey="name"
                            placeholder="Select intent"
                            valuekey="value"
                            values="[]"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class="divider"
                      data-draggable="true"
                    >
                      above
                    </div>
                    <div
                      style="width: 34px;"
                    >
                      <div
                        class="ele"
                      >
                        <div
                          class="wrapper small"
                        >
                          <input
                            data-testid="confidence"
                            name="confidence"
                            placeholder=".9"
                            value=""
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="categorize_as"
                    data-draggable="true"
                  >
                    categorize as
                  </div>
                  <div
                    class="category"
                  >
                    <div
                      class="ele"
                    >
                      <div
                        class="wrapper small"
                      >
                        <input
                          data-testid="Exit Name"
                          name="Exit Name"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                  <span
                    class="fe-x remove_icon"
                    data-testid="remove-case-42ebf1d0-a43e-4ba6-9577-c863b20c88c4"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <div
            class="ele"
          >
            <div
              class="label"
            >
              Result Name
            </div>
            <div
              class="wrapper normal"
            >
              <input
                data-testid="Result Name"
                name="Result Name"
                value="Initial Result Name"
              />
            </div>
            <div
              class="help_text"
            >
              By naming the result, you can reference it later using @results.initial_result_name
               
            </div>
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ClassifyRouterForm should render: default render 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="tabs"
      >
        <div
          class="tab "
        >
          Classifier Input
           
          <span
            class="tab_icon fe-check"
          />
        </div>
      </div>
      <div
        class="header split_by_intent"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Split by Intent
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                errors="[]"
                expressions=""
                name="type"
                namekey="description"
                searchable="true"
                valuekey="type"
                values="[{\\"type\\":\\"split_by_intent\\",\\"name\\":\\"Split by Intent\\",\\"description\\":\\"Split by intent\\",\\"localizeableKeys\\":[\\"exits\\"],\\"aliases\\":[\\"call_classifier\\"],\\"visibility\\":[\\"messaging\\",\\"messaging_background\\",\\"voice\\"],\\"filter\\":\\"classifier\\"}]"
              />
            </div>
          </div>
        </div>
        <p>
          <span>
            Run 
          </span>
          <span
            class="link"
          >
            @input
          </span>
          <span>
             through the classifier...
          </span>
        </p>
        <div
          class="ele"
        >
          <div
            class="normal "
          >
            <temba-select
              cachekey="0"
              data-testid="temba_select_classifier"
              errors="[]"
              expressions=""
              name="Classifier"
              namekey="name"
              placeholder="Select the classifier to use"
              valuekey="uuid"
              values="[{\\"uuid\\":\\"purrington\\",\\"name\\":\\"Purrington\\"}]"
            />
          </div>
        </div>
        <div
          class="case_list_container "
          tabindex="0"
        >
          <div
            class="case_list"
          >
            <div
              class="kase case_list_case"
            >
              <div
                class="ele group"
              >
                <div
                  class="kase has_top_intent"
                  data-draggable="true"
                >
                  <span
                    class="fe-chevrons-expand dnd_icon"
                    data-draggable="true"
                  />
                  <div
                    class="choice"
                  >
                    <div
                      class="small "
                    >
                      <temba-select
                        data-testid="temba_select_operator"
                        errors="[]"
                        expressions=""
                        name="operator"
                        namekey="verboseName"
                        valuekey="type"
                        values="[{\\"type\\":\\"has_top_intent\\",\\"verboseName\\":\\"has top intent\\",\\"operands\\":2,\\"visibility\\":[\\"messaging\\",\\"messaging_background\\",\\"voice\\"]}]"
                      />
                    </div>
                  </div>
                  <div
                    class="multi_operand"
                  >
                    <div
                      style="width: 114px;"
                    >
                      <div
                        class="ele"
                      >
                        <div
                          class="small "
                        >
                          <temba-select
                            data-testid="temba_select_intent"
                            errors="[]"
                            expressions=""
                            hideerrors="true"
                            name="Intent"
                            namekey="name"
                            placeholder="Select intent"
                            valuekey="value"
                            values="[]"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class="divider"
                      data-draggable="true"
                    >
                      above
                    </div>
                    <div
                      style="width: 34px;"
                    >
                      <div
                        class="ele"
                      >
                        <div
                          class="wrapper small"
                        >
                          <input
                            data-testid="confidence"
                            name="confidence"
                            placeholder=".9"
                            value=""
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="categorize_as"
                    data-draggable="true"
                  >
                    categorize as
                  </div>
                  <div
                    class="category"
                  >
                    <div
                      class="ele"
                    >
                      <div
                        class="wrapper small"
                      >
                        <input
                          data-testid="Exit Name"
                          name="Exit Name"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                  <span
                    class="fe-x remove_icon"
                    data-testid="remove-case-2dc85899-0577-4f1b-a620-f12094e34b5e"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <div
            class="ele"
          >
            <div
              class="label"
            >
              Result Name
            </div>
            <div
              class="wrapper normal"
            >
              <input
                data-testid="Result Name"
                name="Result Name"
                value="Result"
              />
            </div>
            <div
              class="help_text"
            >
              By naming the result, you can reference it later using @results.result
               
            </div>
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ClassifyRouterForm updates the result name: updated result name 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="tabs"
      >
        <div
          class="tab "
        >
          Classifier Input
           
          <span
            class="tab_icon fe-check"
          />
        </div>
      </div>
      <div
        class="header split_by_intent"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Split by Intent
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                errors="[]"
                expressions=""
                name="type"
                namekey="description"
                searchable="true"
                valuekey="type"
                values="[{\\"type\\":\\"split_by_intent\\",\\"name\\":\\"Split by Intent\\",\\"description\\":\\"Split by intent\\",\\"localizeableKeys\\":[\\"exits\\"],\\"aliases\\":[\\"call_classifier\\"],\\"visibility\\":[\\"messaging\\",\\"messaging_background\\",\\"voice\\"],\\"filter\\":\\"classifier\\"}]"
              />
            </div>
          </div>
        </div>
        <p>
          <span>
            Run 
          </span>
          <span
            class="link"
          >
            @input
          </span>
          <span>
             through the classifier...
          </span>
        </p>
        <div
          class="ele"
        >
          <div
            class="normal "
          >
            <temba-select
              cachekey="0"
              data-testid="temba_select_classifier"
              errors="[]"
              expressions=""
              name="Classifier"
              namekey="name"
              placeholder="Select the classifier to use"
              valuekey="uuid"
              values="[{\\"uuid\\":\\"purrington\\",\\"name\\":\\"Purrington\\"}]"
            />
          </div>
        </div>
        <div
          class="case_list_container "
          tabindex="0"
        >
          <div
            class="case_list"
          >
            <div
              class="kase case_list_case"
            >
              <div
                class="ele group"
              >
                <div
                  class="kase has_top_intent"
                  data-draggable="true"
                >
                  <span
                    class="fe-chevrons-expand dnd_icon"
                    data-draggable="true"
                  />
                  <div
                    class="choice"
                  >
                    <div
                      class="small "
                    >
                      <temba-select
                        data-testid="temba_select_operator"
                        errors="[]"
                        expressions=""
                        name="operator"
                        namekey="verboseName"
                        valuekey="type"
                        values="[{\\"type\\":\\"has_top_intent\\",\\"verboseName\\":\\"has top intent\\",\\"operands\\":2,\\"visibility\\":[\\"messaging\\",\\"messaging_background\\",\\"voice\\"]}]"
                      />
                    </div>
                  </div>
                  <div
                    class="multi_operand"
                  >
                    <div
                      style="width: 114px;"
                    >
                      <div
                        class="ele"
                      >
                        <div
                          class="small "
                        >
                          <temba-select
                            data-testid="temba_select_intent"
                            errors="[]"
                            expressions=""
                            hideerrors="true"
                            name="Intent"
                            namekey="name"
                            placeholder="Select intent"
                            valuekey="value"
                            values="[]"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class="divider"
                      data-draggable="true"
                    >
                      above
                    </div>
                    <div
                      style="width: 34px;"
                    >
                      <div
                        class="ele"
                      >
                        <div
                          class="wrapper small"
                        >
                          <input
                            data-testid="confidence"
                            name="confidence"
                            placeholder=".9"
                            value=""
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="categorize_as"
                    data-draggable="true"
                  >
                    categorize as
                  </div>
                  <div
                    class="category"
                  >
                    <div
                      class="ele"
                    >
                      <div
                        class="wrapper small"
                      >
                        <input
                          data-testid="Exit Name"
                          name="Exit Name"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                  <span
                    class="fe-x remove_icon"
                    data-testid="remove-case-70ac6ea6-803a-4c33-81c7-dc26803c313f"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <div
            class="ele"
          >
            <div
              class="label"
            >
              Result Name
            </div>
            <div
              class="wrapper normal"
            >
              <input
                data-testid="Result Name"
                name="Result Name"
                value="Updated Result Name"
              />
            </div>
            <div
              class="help_text"
            >
              By naming the result, you can reference it later using @results.updated_result_name
               
            </div>
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
